crates/wysiwyg/src/dom/find_range.rs

```diff
@@ -108,166 +80,265 @@ enum RangeLocation {
     C: Clone,
 {
     // TODO: consider whether cloning DomHandles is damaging performance,
     // and look for ways to pass around references, maybe.
 
+    let mut offset = 0;
+    let locations = do_find_pos(dom, node_handle, start, end, &mut offset);
+
+    if locations.is_empty() {
+        FindResult::NotFound
+    } else {
+        FindResult::Found(locations)
+    }
+}
+
+fn do_find_pos<C>(
+    dom: &Dom<C>,
+    node_handle: DomHandle,
+    start: usize,
+    end: usize,
+    offset: &mut usize,
+) -> Vec<DomLocation>
+where
+    C: Clone,
+{
     let node = dom.lookup_node(node_handle.clone());
+    let mut locations = Vec::new();
+    if *offset > end {
+        return locations;
+    }
     match node {
         DomNode::Text(n) => {
-            let len = n.data().len();
-            if (location == RangeLocation::Start && offset < len)
-                || (location == RangeLocation::End && offset <= len)
-            {
-                FindResult::Found {
-                    node_handle,
-                    offset,
-                }
-            } else {
-                FindResult::NotFound {
-                    new_offset: offset - len,
-                }
+            if let Some(location) = process_text_node(n, start, end, offset) {
+                locations.push(location);
             }
         }
         DomNode::Container(n) => {
-            process_container_node(dom, n, offset, location)
+            locations
+                .extend(process_container_node(dom, n, start, end, offset));
         }
     }
+    locations
 }
 
 fn process_container_node<C: Clone>(
     dom: &Dom<C>,
     node: &ContainerNode<C>,
-    offset: usize,
-    location: RangeLocation,
-) -> FindResult {
-    let mut off = offset;
+    start: usize,
+    end: usize,
+    offset: &mut usize,
+) -> Vec<DomLocation> {
+    let mut results = Vec::new();
     for child in node.children() {
+        let cur_offset = *offset;
         let child_handle = child.handle();
         assert!(!child_handle.is_root(), "Incorrect child handle!");
-        let find_child = find_pos(dom, child_handle, off, location);
-        match find_child {
-            FindResult::Found { .. } => {
-                return find_child;
-            }
-            FindResult::NotFound { new_offset } => {
-                off = new_offset;
-            }
+        let locations = do_find_pos(dom, child_handle, start, end, offset);
+        if !locations.is_empty() {
+            results.extend(locations);
+        }
+        // Container node is completely selected
+        let container_node_len = node.text_len();
+        if !node.handle().is_root()
+            && start <= cur_offset
+            && cur_offset + container_node_len <= end
+        {
+            results.push(DomLocation {
+                node_handle: node.handle(),
+                start_offset: 0,
+                end_offset: container_node_len,
+                location_type: RangeLocationType::Middle,
```

Comment:
I'm not too sure about how to handle `Start`, `End`. Maybe manually mutate the first and last returned locations in `find_range`?